selectrum.el

```diff
@@ -1125,20 +1156,37 @@ HIST, DEF, and INHERIT-INPUT-METHOD, see `completing-read'."
 
 ;;;###autoload
 (defun selectrum-completing-read-multiple
-    (prompt table &optional
-            predicate require-match initial-input
-            hist def inherit-input-method)
+    (prompt table &optional predicate require-match initial-input
+            hist def _inherit-input-method)
   "Read one or more choices using Selectrum.
 Replaces `completing-read-multiple'. For PROMPT, TABLE,
 PREDICATE, REQUIRE-MATCH, INITIAL-INPUT, HIST, DEF, and
 INHERIT-INPUT-METHOD, see `completing-read-multiple'."
-  (ignore initial-input inherit-input-method)
-  (selectrum-read
-   prompt (selectrum--normalize-collection table predicate)
-   :default-candidate (or (car-safe def) def)
-   :require-match require-match
-   :history hist
-   :multiple t))
+  (let* ((crm-completion-table table)
+         (coll (all-completions "" #'crm--collection-fn predicate))
+         (candidates
+          (lambda (input)
+            (let ((beg 0)
+                  (inputs ()))
+              (while (string-match crm-separator input beg)
+                (push (substring input beg (match-beginning 0))
+                      inputs)
+                (setq beg (match-end 0)))
+              (let ((coll (cl-delete-if
+                           (lambda (i)
+                             (member i inputs))
+                           (copy-sequence coll)))
```

Comment:
Would `cl-remove-if` do the same thing faster?

Reply:
The problem is when I use:

    (cl-remove-if
     (lambda (i)
       (member i inputs))
     coll)

and try to input multiple candidates the sequence gets corrupted somehow, I'm don't understand why this is. To test this use it with the definition above and call `describe-face`. Look at the number of candidates and afterwards add one face with `TAB ,` now look at the number of candidates again: they shrinked in size more than one.

Reply:
This is probably because `cl-remove-if` is special-cased to return the same object if there is nothing to remove. Then something downstream destructively modifies the collection.

I looked into this and found that the docstring of `selectrum-read` was not really clear on whether the collection might be modified. I fixed this problem. With the latest commit, I think using the code you have right now (`cl-delete-if` plus `copy-sequence`) should work properly and should be the right thing to do. Let me know.

Reply:
> I looked into this and found that the docstring of selectrum-read was not really clear on whether the collection might be modified. I fixed this problem

If the collection is a function it may still be later be modified by the sorting function which is why I need to copy the sequence above every time it is called. `copy-sequence` is pretty fast maybe we could make it the default and also do it before any sorting?

Reply:
The problem is that `copy-sequence` is fast, but it generates a lot of garbage due to duplicating the collection, and garbage collection _really_ kills latency. What's slow in Elisp is really not obvious at first glance.

> the candidates returned by the function above can be modified and you can't just reuse the value inside the closure

If we don't copy the candidates returned from a function collection when MAY-MODIFY-CANDIDATES is nil, that's a bug.

Reply:
> The problem is that copy-sequence is fast, but it generates a lot of garbage due to duplicating the collection, and garbage collection really kills latency. What's slow in Elisp is really not obvious at first glance.

Ah, yes thanks for pointing that out, totally forgot about gc!

> If we don't copy the candidates returned from a function collection when MAY-MODIFY-CANDIDATES is nil, that's a bug.

I will open an issue for that later.